src/vos/vos_obj_index.c

```diff
@@ -111,30 +118,47 @@ oi_rec_alloc(struct btr_instance *tins, d_iov_t *key_iov,
 	return 0;
 }
 
+struct oi_delete_arg {
+	void		*cont;
+	uint32_t	only_delete_entry:1;
+};
+
 static int
 oi_rec_free(struct btr_instance *tins, struct btr_record *rec, void *args)
 {
 	struct umem_instance	*umm = &tins->ti_umm;
 	struct vos_obj_df	*obj;
 	struct ilog_desc_cbs	 cbs;
-	daos_handle_t		 coh;
+	struct oi_delete_arg	*del_arg = args;
+	daos_handle_t		 coh = { 0 };
 	int			 rc;
 
 	obj = umem_off2ptr(umm, rec->rec_off);
 
-	vos_ilog_desc_cbs_init(&cbs, tins->ti_coh);
-	rc = ilog_destroy(umm, &cbs, &obj->vo_ilog);
-	if (rc != 0) {
-		D_ERROR("Failed to destroy incarnation log: "DF_RC"\n",
-			DP_RC(rc));
-		return rc;
-	}
+	/* Normally it should delete both ilog and vo_tree, but during upgrade
+	 * the new OID (with new layout version) will share the same ilog and
+	 * vos_tree with the old OID (with old layout version), so it will only
+	 * delete the entry in this case.
+	 */
+	if (del_arg != NULL && del_arg->only_delete_entry) {
+		memset(&obj->vo_ilog, 0, sizeof(obj->vo_ilog));
```

Comment:
I'm confused by this change, I will take up up offline.   But first comment is that the ilog may allocate new space internally so memset could result in storage leak.

Reply:
During upgrade,  the new OI entry does not need to recreate the ilog or vos_tree, instead they will share the same ilog and vos_tree with the old entry, and these old entry fee will only reset the ilog and vos_tree.